README.md
# clojure-reducing

FIXME: description

## Installation

Download from http://example.com/FIXME.

## Usage

FIXME: explanation

    $ java -jar clojure-reducing-0.1.0-standalone.jar [args]

## Options

FIXME: listing of options this app accepts.

## Examples

...

### Bugs

...

### Any Other Sections
### That You Think
### Might be Useful

## License

Copyright © 2021 FIXME

This program and the accompanying materials are made available under the
terms of the Eclipse Public License 2.0 which is available at
http://www.eclipse.org/legal/epl-2.0.

This Source Code may also be made available under the following Secondary
Licenses when the conditions for such availability set forth in the Eclipse
Public License, v. 2.0 are satisfied: GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or (at your
option) any later version, with the GNU Classpath Exception which is available
at https://www.gnu.org/software/classpath/license.html.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2021-01-01
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2021-01-01
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/clojure-reducing/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/clojure-reducing/compare/0.1.0...0.1.1


test/clojure_reducing/core_test.clj
(ns clojure-reducing.core-test
  (:require [clojure.test :refer :all]
            [clojure-reducing.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


src/clojure_reducing/core.clj
(ns clojure-reducing.core
  (:require
   [clojure.java.io :as io]
   [clojure.data.csv :as csv]
   [semantic-csv.core :as sc])
  (:gen-class))

(defn -main
  "I don't do a whole lot ... yet."
  [& args]
  (println "Hello, World!"))

;; Simple reducing
(def simple-reducing (reduce + [8 4000 10 300]))

;; Internally the function + when called with more than
;; 2 arguments uses a version of reduce  
(def simple-reducing-2 (apply + [8 4000 10 300]))

;; ============================================
;; Finding the Day with the Maximum Temperature

(def weather-days
  [{:max 31
    :min 27
    :description :sunny
    :date "2019-09-24"}
   {:max 28
    :min 25
    :description :cloudy
    :date "2019-09-25"}
   {:max 22
    :min 18
    :description :rainy
    :date "2019-09-26"}
   {:max 23
    :min 16
    :description :stormy
    :date "2019-09-27"}
   {:max 35
    :min 19
    :description :sunny
    :date "2019-09-28"}])

;; We need to be able to write functions that return the entire map for the day 
;; with the highest maximum temperature, the lowest minimum, and so on

(defn max-temp []
  (reduce (fn [max-day-so-far this-day]
            (if (> (:max this-day) (:max max-day-so-far))
              this-day
              max-day-so-far))
          weather-days))

(defn min-max-temp []
  (reduce (fn [min-max-day-so-far this-day]
            (if (< (:max this-day) (:max min-max-day-so-far))
              this-day
              min-max-day-so-far))
          weather-days))

;; ============================================

(defn min-and-max []
  (reduce (fn [{:keys [minimum maximum]} new-number]
            {:minimum (if (and minimum (> new-number minimum))
                        minimum
                        new-number)
             :maximum (if (and maximum (< new-number maximum))
                        maximum
                        new-number)})
          {} ;; Init value
          [5 23 5004 845 22]))

;; Partitioning
;; (partition 3 [1 2 3 4 5 6 7 8 9 10])
;; (partition-all 3 [1 2 3 4 5 6 7 8 9 10])

(defn segment-by-sum [limit ns]
  (let [result (reduce (fn [{:keys [segments current] :as acc} n]
                         (let [current-with-n (conj current n)
                               total-with-n (apply + current-with-n)]
                           (if (> total-with-n limit)
                             (assoc acc
                                    :segments (conj segments current)
                                    :current [n])
                             (assoc acc :current current-with-n))))
                       {:segments [] :current []}
                       ns)]
    (conj (:segments result) (:current result))))

;; =============================================================================

(def numbers [4 9 2 3 7 9 5 2 6 1 4 6 2 3 3 6 1])

(defn parity-totals [ns]
  (:ret
   (reduce (fn [{:keys [current] :as acc} n]
             (if (and (seq current)
                      (or (and (odd? (last current)) (odd? n))
                          (and (even? (last current)) (even? n))))
               (-> acc
                   (update :ret conj [n (apply + current)])
                   (update :current conj n))
               (-> acc
                   (update :ret conj [n 0])
                   (assoc :current [n]))))
           {:current [] :ret []}
           ns)))

;; Measuring Elevation Differences on Slopes
;; =========================================

(def distance-elevation
  [[0 400]
   [12.5 457]
   [19 622]
   [21.5 592]
   [29 615]
   [35.5 892]
   [39 1083]
   [43 1477]
   [48.5 1151]
   [52.5 999]
   [57.5 800]
   [62.5 730]
   [65 1045]
   [68.5 1390]
   [70.5 1433]
   [75 1211]
   [78.5 917]
   [82.5 744]
   [84 667]
   [88.5 860]
   [96 671]
   [99 584]
   [108 402]
   [115.5 473]])

(defn same-slope-as-current? [current elevation]
  (or (= 1 (count current))
      (let [[[_ next-to-last] [_ the-last]] (take-last 2 current)]
        (or (>= next-to-last the-last elevation)
            (<= next-to-last the-last elevation)))))

(defn distances-elevation-to-next-peak-or-valley
  [data]
  (->
   (reduce (fn [{:keys [current] :as acc}
                [distance elevation :as this-position]]
             (cond (empty? current)
                   {:current [this-position]
                    :calculated [{:race-position distance
                                  :elevation elevation
                                  :distance-to-next 0
                                  :elevation-to-next 0}]}
                   (same-slope-as-current? current elevation)
                   (-> acc
                       (update :current conj this-position)
                       (update :calculated conj
                               {:race-position distance
                                :elevation elevation
                                :distance-to-next (- (first (first current)) distance)
                                :elevation-to-next (- (second (first current)) elevation)}))
                   :else
                   (let [[prev-distance
                          prev-elevation
                          :as peak-or-valley] (last current)]
                     (-> acc
                         (assoc :current [peak-or-valley this-position])
                         (update :calculated conj
                                 {:race-position distance
                                  :elevation elevation
                                  :distance-to-next (- prev-distance distance)
                                  :elevation-to-next (- prev-elevation elevation)})))))
           {:current []
            :calculated []}
           (reverse data))
   :calculated
   reverse))

;; =============================================================================
;; Reducing without reduce

(def matches
  [{:winner-name "Kvitova P."
    :loser-name "Ostapenko J."
    :tournament "US Open"
    :location "New York"
    :date "2016-08-29"}
   {:winner-name "Kvitova P."
    :loser-name "Buyukakcay C."
    :tournament "US Open"
    :location "New York"
    :date "2016-08-31"}
   {:winner-name "Kvitova P."
    :loser-name "Svitolina E."
    :tournament "US Open"
    :location "New York"
    :date "2016-09-02"}
   {:winner-name "Kerber A."
    :loser-name "Kvitova P."
    :tournament "US Open"
    :location "New York"
    :date "2016-09-05"}
   {:winner-name "Kvitova P."
    :loser-name "Brengle M."
    :tournament "Toray Pan Pacific Open"
    :location "Tokyo"
    :date "2016-09-20"}
   {:winner-name "Puig M."
    :loser-name "Kvitova P."
    :tournament "Toray Pan Pacific Open"
    :location "Tokyo"
    :date "2016-09-21"}])

;; Lookup table
;; to get values simply use (get matches-by-date "The-date")
(def matches-by-date (zipmap (map :date matches) matches))

(def dishes
  [{:name "Carrot Cake"
    :course :dessert}
   {:name "French Fries"
    :course :main}
   {:name "Celery"
    :course :appetizer}
   {:name "Salmon"
    :course :main}
   {:name "Rice"
    :course :main}
   {:name "Ice Cream"
    :course :dessert}])

;; Group data by :course
;; (group-by :course dishes)
;; =============================================================================
;; 
;; Quick Summary Statistics with group-by

(def csv "resources/match_scores_1968-1990_unindexed_csv.csv")

(defn tennis-csv->tournament-match-counts [csv]
  (with-open [r (io/reader csv)]
    (->> (csv/read-csv r)
         sc/mappify
         (group-by :tourney_slug)
         (map (fn [[k ms]] [k (count ms)]))
         (into {}))))

(def tournaments-totals (tennis-csv->tournament-match-counts csv))

;; =============================================================================
;;
;; Complex Accumulation with reduce

(defn win-loss-by-player [csv]
  (with-open [r (io/reader csv)]
    (->> (csv/read-csv r)
         sc/mappify
         (reduce (fn [acc {:keys [winner_slug loser_slug]}]
                   (-> acc
                       (update-in [winner_slug :wins] 
                                  (fn [wins] (inc (or wins 0))))
                       (update-in [loser_slug :losses]
                                  (fn [losses] (inc (or losses 0))))))
                 {}))))


src/clojure_reducing/tennis_reduce.clj
(ns tennis-reduce)

(def serena-williams-2015
  [{:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Van Uytvanck A.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-20"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Zvonareva V.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-22"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Svitolina E.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-24"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Muguruza G.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-26"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Cibulkova D.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-28"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Keys M.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-29"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Sharapova M.",
    :tournament "Australian Open",
    :location "Melbourne",
    :date "2015-01-31"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Niculescu M.",
    :tournament "BNP Paribas Open",
    :location "Indian Wells",
    :date "2015-03-14"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Diyas Z.",
    :tournament "BNP Paribas Open",
    :location "Indian Wells",
    :date "2015-03-15"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Stephens S.",
    :tournament "BNP Paribas Open",
    :location "Indian Wells",
    :date "2015-03-17"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Bacsinszky T.",
    :tournament "BNP Paribas Open",
    :location "Indian Wells",
    :date "2015-03-19"}
   {:loser-sets-won nil,
    :winner-sets-won nil,
    :winner-name "Halep S.",
    :loser-name "Williams S.",
    :tournament "BNP Paribas Open",
    :location "Indian Wells",
    :date "2015-03-21"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Niculescu M.",
    :tournament "Sony Ericsson Open",
    :location "Miami",
    :date "2015-03-28"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Bellis C.",
    :tournament "Sony Ericsson Open",
    :location "Miami",
    :date "2015-03-29"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Kuznetsova S.",
    :tournament "Sony Ericsson Open",
    :location "Miami",
    :date "2015-03-30"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Lisicki S.",
    :tournament "Sony Ericsson Open",
    :location "Miami",
    :date "2015-04-01"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Halep S.",
    :tournament "Sony Ericsson Open",
    :location "Miami",
    :date "2015-04-03"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Suarez Navarro C.",
    :tournament "Sony Ericsson Open",
    :location "Miami",
    :date "2015-04-04"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Brengle M.",
    :tournament "Mutua Madrid Open",
    :location "Madrid",
    :date "2015-05-03"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Stephens S.",
    :tournament "Mutua Madrid Open",
    :location "Madrid",
    :date "2015-05-04"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Azarenka V.",
    :tournament "Mutua Madrid Open",
    :location "Madrid",
    :date "2015-05-06"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Suarez Navarro C.",
    :tournament "Mutua Madrid Open",
    :location "Madrid",
    :date "2015-05-07"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Kvitova P.",
    :loser-name "Williams S.",
    :tournament "Mutua Madrid Open",
    :location "Madrid",
    :date "2015-05-08"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Pavlyuchenkova A.",
    :tournament "Internazionali BNL d'Italia",
    :location "Rome",
    :date "2015-05-12"}
   {:loser-sets-won nil,
    :winner-sets-won nil,
    :winner-name "Mchale C.",
    :loser-name "Williams S.",
    :tournament "Internazionali BNL d'Italia",
    :location "Rome",
    :date "2015-05-14"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Hlavackova A.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-05-26"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Friedsam A.L.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-05-28"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Azarenka V.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-05-30"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Stephens S.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-06-01"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Errani S.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-06-03"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Bacsinszky T.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-06-04"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Safarova L.",
    :tournament "French Open",
    :location "Paris",
    :date "2015-06-06"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Gasparyan M.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-06-29"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Babos T.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-07-01"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Watson H.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-07-03"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Williams V.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-07-06"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Azarenka V.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-07-07"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Sharapova M.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-07-09"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Muguruza G.",
    :tournament "Wimbledon",
    :location "London",
    :date "2015-07-11"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Bonaventure Y.",
    :tournament "Collector Swedish Open",
    :location "Bastad",
    :date "2015-07-15"}
   {:loser-sets-won nil,
    :winner-sets-won nil,
    :winner-name "Koukalova K.",
    :loser-name "Williams S.",
    :tournament "Collector Swedish Open",
    :location "Bastad",
    :date "2015-07-16"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Pennetta F.",
    :tournament "Rogers Cup",
    :location "Toronto",
    :date "2015-08-11"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Petkovic A.",
    :tournament "Rogers Cup",
    :location "Toronto",
    :date "2015-08-14"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Vinci R.",
    :tournament "Rogers Cup",
    :location "Toronto",
    :date "2015-08-15"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Bencic B.",
    :loser-name "Williams S.",
    :tournament "Rogers Cup",
    :location "Toronto",
    :date "2015-08-15"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Pironkova T.",
    :tournament "Western & Southern Financial Group Women's Open",
    :location "Cincinnati",
    :date "2015-08-19"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Knapp K.",
    :tournament "Western & Southern Financial Group Women's Open",
    :location "Cincinnati",
    :date "2015-08-20"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Ivanovic A.",
    :tournament "Western & Southern Financial Group Women's Open",
    :location "Cincinnati",
    :date "2015-08-21"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Svitolina E.",
    :tournament "Western & Southern Financial Group Women's Open",
    :location "Cincinnati",
    :date "2015-08-23"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Halep S.",
    :tournament "Western & Southern Financial Group Women's Open",
    :location "Cincinnati",
    :date "2015-08-23"}
   {:loser-sets-won 0,
    :winner-sets-won 1,
    :winner-name "Williams S.",
    :loser-name "Diatchenko V.",
    :tournament "US Open",
    :location "New York",
    :date "2015-09-01"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Bertens K.",
    :tournament "US Open",
    :location "New York",
    :date "2015-09-02"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Mattek-Sands B.",
    :tournament "US Open",
    :location "New York",
    :date "2015-09-05"}
   {:loser-sets-won 0,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Keys M.",
    :tournament "US Open",
    :location "New York",
    :date "2015-09-06"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Williams S.",
    :loser-name "Williams V.",
    :tournament "US Open",
    :location "New York",
    :date "2015-09-09"}
   {:loser-sets-won 1,
    :winner-sets-won 2,
    :winner-name "Vinci R.",
    :loser-name "Williams S.",
    :tournament "US Open",
    :location "New York",
    :date "2015-09-11"}])

(defn streak-string [current-wins current-losses]
  (cond (pos? current-wins) (str "Won " current-wins)
        (pos? current-losses) (str "Lost " current-losses)
        :else "First match of the year"))

(defn serena-williams-win-loss-streaks [matches]
  (:matches 
   (reduce (fn [{:keys [current-wins current-losses] :as acc}
                {:keys [winner-name] :as match}]
             (let [this-match (assoc match 
                                     :current-streak 
                                     (streak-string current-wins current-losses))
                   serena-vistory? (= winner-name "Williams S.")]
               (-> acc
                   (update :matches #(conj % this-match))
                   (assoc :current-wins (if serena-vistory? (inc current-wins) 0))
                   (assoc :current-losses (if serena-vistory? 0 (inc current-losses))))))
          {:matches []
           :current-wins 0
           :current-losses 0}
          matches)))